### PR TITLE
option to disable metadata and graphql apis (close #1088)

### LINKED ISSUE #1088: Add option to disable metadata endpoints
Hasura has 2 additional endpoints that are used to manage the database, create migrations, etc.  They are used by the console and cli. 

Currently there is no option to disable those endpoints.  Leaving full access to the database in cases when the access-key gets compromised.

My suggestion is to add an option to disable those metadata endpoints.  Where sys-admins can enable it to make modifications/migrations in the database and disable it when the system goes live.

Currently I am manually disabling it in my proxy.

Thanks

### PR DESCRIPTION
**Pull request by nizar-m**
### Description

Currently both the Metadata (/api/1/table and /v1/query) and the GraphQL endpoints (/v1alpha1/graphql) are enabled by default.

A new flag `--enabled-apis` (EnvVar:  HASURA_GRAPHQL_ENABLED_APIS) is introduced. 
 - Accepts comma list of APIs to be enabled
   -  `--enabled-apis metadata` enables Metadata 
   -  `--enabled-apis graphql` enabled GraphQL
   -  `--enabled-apis metadata,graphql` enables both
 -  By default, both Metadata and GraphQL APIs will be enabled

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [ ] Console
- [ ] CLI
- [x] Docs
- [ ] Community Content
- [ ] Build System

### Related Issue

#1088 

### Solution and Design

- Parse EnvVar `HASURA_GRAPHQL_ENABLED_APIS` as a comma separated list of APIs to be enabled
- Parse flag `--enabled-apis`
- Disable /v1/query, /v1/template, /api/1/table, and console if RQL is disabled 
- Disable /v1alpha1/graphql if GraphQL is disabled

### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Docs update
- [ ] Community content

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation. 
- [ ] I have updated the documentation accordingly.
- [x] I have added required tests.

### CODE REVIEW

File: server/src-lib/Hasura/Server/App.hs [hunk 1]
```diff
@@ -318,41 +327,45 @@ httpApp corsCfg serverCtx enableConsole enableTelemetry = do
...
+        expQuery <- parseBody
+        gqlExplainHandler expQuery
+
+      post "v1alpha1/graphql" $ mkSpockAction GH.encodeGQErr serverCtx $ do
+        query <- parseBody
+        v1Alpha1GQHandler query
+
+        -- get "v1alpha1/graphql/schema" $
+        --   mkSpockAction encodeQErr serverCtx v1Alpha1GQSchemaHandler
+    when enableRQL $  do
```
Comment (0x777) [inline server/src-lib/Hasura/Server/App.hs]: Let's put this under the earlier `when enableRQL` block

**Comment by 0x777** [comment]
@nizar-m  After a bit of discussion, we want to call these `apis` instead of interfaces and the `RQL` interface as `metadata`. So, the env variable would be `HASURA_GRAPHQL_ALLOWED_APIS`. By default it is, `HASURA_GRAPHQL_ALLOWED_APIS=metadata,graphql`. Also regarding the command line flag, should we just call it `--allowed-apis` which will accept comma separated values? What is the typical practice? @shahidhk @rikinsk

**Comment by shahidhk** [comment]
`HASURA_GRAPHQL_ALLOWED_APIS=metadata,graphql` as env var and `--allowed-apis=metadata,graphl` as flag is recommended.

**Comment by shahidhk** [comment]
I'd also suggest changing `ALLOWED` to`ENABLED` in-lieu with `ENABLE_CONSOLE`. What rings to me when using `ALLOWED` is `ALLOWED_ROLES`

So, `HASURA_GRAPHQL_ENABLED_APIS=metadata,graphql` as env var and `--enabled-apis=metadata,graphl` as flag.

File: server/src-lib/Hasura/Server/Init.hs [hunk 2]
```diff
@@ -584,6 +605,13 @@ readHookType tyS =
     "POST" -> Right AHTPost
     _      -> Left "Only expecting GET / POST"
 
+readAPIs :: String -> Either String [API]
+readAPIs = mapM readAPI . T.splitOn "," . T.pack
+  where readAPI si = case T.toUpper $ T.strip si of
+          "RQL"     -> Right RQL
```
Comment (shahidhk) [inline server/src-lib/Hasura/Server/Init.hs]: We need to change `RQL` to `METADATA`

File: docs/graphql/manual/deployment/graphql-engine-flags/reference.rst [hunk 3]
```diff
@@ -139,6 +139,9 @@ For ``serve`` sub-command these are the flags and ENV variables available:
      - ``HASURA_GRAPHQL_TX_ISOLATION``
      - transaction isolation. read-committed / repeatable-read / serializable (default: read-commited)
 
+   * - ``--enabled-apis <APIS>``
+     - ``HASURA_GRAPHQL_ENABLED_APIS``
+     - Comma separated list of APIs to be enabled. (default: ``metadata,graphql``)
```
Comment (rikinsk) [inline docs/graphql/manual/deployment/graphql-engine-flags/reference.rst]: Can we add all possible API values here

**Pull request merged**